src/dmd/errors.d

```diff
@@ -649,15 +649,26 @@ private void _vdeprecationSupplemental(const ref Loc loc, const(char)* format, v
 }
 
 /**
- * Call this after printing out fatal error messages to clean up and exit
- * the compiler.
+ * The type of the fatal error handler
+ * Returns: true if error handling is done, false to do exit(EXIT_FAILURE)
+ */
+alias FatalErrorHandler = bool delegate();
+
+/**
+ * The fatal error handler.
+ * If non-null it will be called for every fatal() call issued by the compiler.
+ */
+__gshared FatalErrorHandler fatalErrorHandler;
+
+/**
+ * Call this after printing out fatal error messages to clean up and exit the
+ * compiler. You can also set a fatalErrorHandler to override this behaviour.
  */
 extern (C++) void fatal()
 {
-    version (none)
-    {
-        halt();
-    }
+    if (fatalErrorHandler && fatalErrorHandler())
+        return;
```

Comment:
The whole point of this is that what is fatal for DMD is not necessarily fatal for an application using DMD as a library, e.g. hitting too many errors: its fatal for DMD because there is no point in issuing a cascade of error (unless requested), for an editor using DMD as a completion engine, terminating because the user typo'd a bunch of stuff makes no sense.

Reply:
Exactly. We shouldn't force the `libdmd` user to `exit(1)`, even when it returns. We should instead fix the issues to not segfault. Currently `fatal()` is being used to fail fast and prevent weird cascade and possibly unrelated errors to show to the end-user while compiling. With `libdmd` we may have better control of that. My initial idea was to control that via the global state, but being able to fine-tune that can be great for other users.

Also, one of the things that some linters/completion engines fail is the fact that, because of some typos, analysis just stops working, and we don't want that in those tools. Well, yeah, you may also have weird errors, but it's better to have some hints rather than nothing due to unrelated changes.